### ISSUE TITLE
Add ability to get package without export

### ISSUE THREAD

**Issue by rconde01** [conan-io/conan#866]
The export directory can have a significant size. The client of a package usually just needs the package directory. It would be nice if you could just get the package.

**Comment by lasote** [comment]
You always need the recipe `conanfile.py` to know what package do you need, it will mix the declared settings with your settings values. The rest of files exported (if there are), could be theoretically downloaded later only for building (if no trick depending on those files are used in configure methods) but would require some important changes in conan.

**Comment by memsharded** [comment]
That is true, it is an optimization, but could make sense for packages snapshoting the sources in them, when the sources are very large. My proposal would be:

- Add a ``exports_sources`` field, that complements the existing ``exports``, that marks the sources, files not necessary for the recipe (note that the recipe itself might have different .py or even .txt files that are needed), but for building from sources
- The existing ``exports`` behavior is not changed
- What is àdded to ``exports_sources`` goes to a different tgz file
- The existing ``get_recipe()`` method keeps downloading the existing tgz as usual
- The existing ``get_recipe()`` method will not download by default the new tgz with the sources
- The ``Installer``, will call again ``get_recipe()`` with a new argument, or another method, calling the same API endpoint, when it is necessary to build from sources, to retrieve the new tgz and unzip it
- It is necessary to be sure that ``upload`` will not be affected, and the new tgz has to be downloaded previous to upload recipes to remotes.

Sounds good?

**Comment by rconde01** [comment]
Unclear if "sounds good was for me or lasote", but it is probably ok. Though, can you give an example of what would be in exports vs exports_sources?

**Comment by rconde01** [comment]
Also - seems like exports and exports_sources should go to different subdirectories...otherwise they could overlap/interfere. Or maybe you already meant that?

**Comment by memsharded** [comment]
"sounds good" is for both, all feedback welcome.

``exports`` remains "as-is", otherwise it would be breaking behaviour. ``exports_sources`` will define those, already existing in ``exports`` that can be treated specially and only be downloaded when necessary to build the package. E.g.

```python
exports = "src/*", "myauxfuncs.py", "CMakeLists.txt"
exports_sources = "src/*", "CMakeLists.txt"
```

No, not necessary to use different subdirectories, they won't interfere, as I proposed, files will go to different tgz files, and they will not be duplicated in them, each file will only be in one tgz. And it is much better to keep files in the current folder, as the change will be transparent to the rest of the codebase. I will have the idea in mind while implementing, just in case, thanks for the suggestion.

**Comment by rconde01** [comment]
ok - I think i'm not being clear:

currently we have:

storage/pkg/x.x.x/me/channel
--export
--build
--source
--package

does this become:
storage/pkg/x.x.x/me/channel
--export
--export_sources
--build
--source
--package

...or something else?

Also, i'm trying to think of what you have in exports besides the conan file and the sources. Your design works regardless, but if there's not a good use case, then I'm wondering if:

conan install pkg/x.x.x@me/channel --min-exports

or something might be simpler. If there's a valid use case, then your design (or something like it) is obviously necessary.

Also, if my description above is correct, you probably want something like:

conan remove pkg/x.x.x@me/channel -es

**Comment by memsharded** [comment]
Yes, the case is that some users are splitting their recipes in multiple files, mainly python, but also text files with data (like the version, which is read by the python recipe). So the ``--min-exports`` can't be a user choice, it has to be defined by the package creator, they have to differentiate which files are sources and which files are part of the recipe.

So the idea is not to change the storage layout at all, it is not necessary, and it will add extra complexity.

**Comment by rconde01** [comment]
ok - splitting the recipe, that makes sense to me.

But if you can not have the sources, because you don't need to build, then you should be able to remove them once you have already built. But since they're in the same directory, that seems complicated. How do you know what to delete?

**Comment by memsharded** [comment]
No, I think is better not to remove them once the package is built, because if the package is to be uploaded, the recipe has to be complete and self-contained. In fact, I think the weakest point of the approach is to ensure that all the files are there before doing uploads, so when no builds from sources have been fired, it will be necessary to force the download of the package recipe sources into the ``exports`` folder, so the upload works.

**Comment by memsharded** [comment]
Released in 0.19. I am writing the docs now.
